### PR TITLE
GH-39583: [C++] Fix the issue of ExecBatchBuilder when appending consecutive tail rows with the same id may exceed buffer boundary (for fixed size types)

### LINKED ISSUE #39583: [C++] Segmentation fault in `ExecBatchBuilder`
### Describe the bug, including details regarding any error messages, version, and platform.

This is a subsequent issue of https://github.com/apache/arrow/issues/32570.

In the last fix https://github.com/apache/arrow/pull/39234, I didn't realize there were similar issue for fixed size types.

I am now able to produce the segmentation fault using fixed size types, listed in the following UT:
```
TEST(ExecBatchBuilder, AppendBatchDupRowsFixedSize) {
  std::unique_ptr<MemoryPool> owned_pool = MemoryPool::CreateDefault();
  MemoryPool* pool = owned_pool.get();
  {
    // 63-byte data occupying almost one minimal 64-byte aligned memory region.
    ExecBatch batch_fsb = JSONToExecBatch({fixed_size_binary(9)}, R"([
        ["000000000"],
        ["000000000"],
        ["000000000"],
        ["000000000"],
        ["000000000"],
        ["000000000"],
        ["123456789"]])");  // 9-byte tail row, not aligned to a word.
    ASSERT_EQ(batch_fsb[0].array()->buffers[1]->capacity(), 64);
    ExecBatchBuilder builder;
    uint16_t row_ids[2] = {6, 6};
    ASSERT_OK(builder.AppendSelected(pool, batch_fsb, 2, row_ids, /*num_cols=*/1));
    ExecBatch built = builder.Flush();
    ExecBatch batch_fsb_appended =
        JSONToExecBatch({fixed_size_binary(9)}, R"([["123456789"], ["123456789"]])");
    ASSERT_EQ(batch_fsb_appended, built);
    ASSERT_NE(0, pool->bytes_allocated());
  }
  ASSERT_EQ(0, pool->bytes_allocated());
}
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

#39583 is a subsequent issue of #32570 (fixed by #39234). The last issue and fixed only resolved var length types. It turns out fixed size types have the same issue.

### What changes are included in this PR?

Do the same fix of #39234 for fixed size types.

### Are these changes tested?

UT included.

### Are there any user-facing changes?

* Closes: #39583

### CODE REVIEW

**Comment by zanmato1984** [comment]
Hi @pitrou , I wonder if you could help to review this PR as this is very similar to its predecessor #39234 reviewed by you. Much appreciate it!

File: cpp/src/arrow/compute/light_array.cc [hunk 1]
```diff
@@ -383,27 +383,22 @@ int ExecBatchBuilder::NumRowsToSkip(const std::shared_ptr<ArrayData>& column,
...
 
   int num_rows_left = num_rows;
   int num_bytes_skipped = 0;
   while (num_rows_left > 0 && num_bytes_skipped < num_tail_bytes_to_skip) {
+    --num_rows_left;
+    int row_id_removed = row_ids[num_rows_left];
     if (column_metadata.is_fixed_length) {
-      if (column_metadata.fixed_length == 0) {
-        num_rows_left = std::max(num_rows_left, 8) - 8;
-        ++num_bytes_skipped;
```
Comment (pitrou) [inline cpp/src/arrow/compute/light_array.cc:393]: This was for boolean data, right? Is it ok to remove this?

Comment (zanmato1984) [reply cpp/src/arrow/compute/light_array.cc:393]: Yes. Fixed size types of length `0/1/2/4/8` are using element-wise copying, i.e., no "word-wise then tail-bytes" copying. They don't go to this method:
https://github.com/apache/arrow/blob/3acc2eab739bfcc89bd39c3242dad154a980acb7/cpp/src/arrow/compute/light_array.cc#L513-L546

So I think we can simplify this method by not specially dealing with boolean type, thus the added `ARROW_DCHECK` several lines above.

**Comment by pitrou** [comment]
@zanmato1984 Can you rebase on git main?

**Comment by zanmato1984** [comment]
> @zanmato1984 Can you rebase on git main?

Sure. Done.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Review by pitrou — APPROVED** [review]
LGTM, thank you @zanmato1984 . The original code quality is quite bad but this improves things slightly.

**Comment by pitrou** [comment]
@raulcd If you do another RC, this would be a good candidate fix to add.

**Pull request merged**